docs/convenience.rst

```diff
@@ -63,119 +61,63 @@ overcome transient failures.  This typically looks something like:
         else:
            break
 
-This is rather ugly.
+This is rather ugly and easy to get wrong.
 
-Transaction managers provide a helper for this case. To show this,
-we'll use a contrived example:
+Transaction managers provide two helpers for this case.
 
-.. doctest::
+Running and retrying functions as transactions
+______________________________________________
 
-    >>> ntry = 0
-    >>> with transaction.manager:
-    ...      dm['ntry'] = 0
-
-    >>> import transaction.interfaces
-    >>> class Retry(transaction.interfaces.TransientError):
-    ...     pass
-
-    >>> for attempt in transaction.manager.attempts():
-    ...     with attempt as t:
-    ...         t.note('test')
-    ...         print("%s %s" % (dm['ntry'], ntry))
-    ...         ntry += 1
-    ...         dm['ntry'] = ntry
-    ...         if ntry % 3:
-    ...             raise Retry(ntry)
-    0 0
-    0 1
-    0 2
-
-The raising of a subclass of TransientError is critical here. It's
-what signals that the transaction should be retried.  It is generally
-up to the data manager to signal that a transaction should try again
-by raising a subclass of TransientError (or TransientError itself, of
-course).
-
-You shouldn't make any assumptions about the object returned by the
-iterator.  (It isn't a transaction or transaction manager, as far as
-you know. :)  If you use the ``as`` keyword in the ``with`` statement,
-a transaction object will be assigned to the variable named.
-
-By default, it tries 3 times. We can tell it how many times to try:
+The first helper runs a function as a transaction::
 
-.. doctest::
+    def do_somthing():
+        "Do something"
+        ... some something ...
 
-    >>> for attempt in transaction.manager.attempts(2):
-    ...     with attempt:
-    ...         ntry += 1
-    ...         if ntry % 3:
-    ...             raise Retry(ntry)
-    Traceback (most recent call last):
-    ...
-    Retry: 5
+    transaction.manager.run(do_somthing)
 
-It it doesn't succeed in that many times, the exception will be
-propagated.
+Of course you can run this as a decorator::
 
-Of course, other errors are propagated directly:
+    @transaction.manager.run
+    def do_somthing():
```

Comment:
I guess my main objection is that it just doesn't feel Pythonic to me (I know that "Pythonic" is a nebulous touchy-feely concept). I too have used this type of pattern extensively in other languages (I wrote Lisp code for 5 years), but it just seems very uncommon in Python. I've heard arguments against Zope and Zope-adjacent libraries that they're just not Pythonic, and I'd hope we can dispel that myth, not fuel it. 

Do you know of some other widely-used Python library/framework that encourages the definition-is-execution approach? I'd gladly drop my objection in that case. 😄